### PR TITLE
Update build process to include integrated CLI

### LINKED ISSUE #159789: Integrate the VS Code CLI with our product builds
Depends on https://github.com/microsoft/vscode/issues/159777

Currently the CLI repo is not OSS, so I suggest we pull it in like we do vscode-distro. The CLI can publish its binaries to an ADO feed, which can then be referenced in the product.json for stable builds

For Insiders builds, we should pull the last version of the CLI before the Insiders commit was made to avoid having to manually bump the product.json each day.

### PR DESCRIPTION
**Pull request by connor4312**
[Review without whitespace diffs](https://github.com/microsoft/vscode/pull/161640/files?w=1)

- Adds an off-by-default option that controls whether to build the Rust CLI
- The "standalone cli" is built first in parallel, and then mixed into the standard builds. The binary name is suffixed with the quality, e.g. `code-tunnel-exploration`
- The build process is largely based on the vscode-cli repo, but with a new `prepare.ts` script that sets variables for product.json and version information
- Adds a compile and unit test run to PR builds
- Updates to use hard tabs instead of spaces as Joao mentioned in the initial PR.

Example pipeline run: https://dev.azure.com/monacotools/Monaco/_build/results?buildId=186450&view=results

Fixes https://github.com/microsoft/vscode/issues/159789

### CODE REVIEW

**Review by joaomoreno — COMMENTED** [review]
Not yet a full review, but just an observation from a chat discussion: we should zip these up before they end up in our storage and further eat at our CDN costs.

> zipping `cli-alpine-x64` makes the asset go from 10MB to 4MB

**Comment by joaomoreno** [comment]
Marking as draft, pending a full review early October.

File: build/azure-pipelines/mixin-distro-win32.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (joaomoreno) [inline build/azure-pipelines/mixin-distro-win32.yml:1]: New line?

File: build/azure-pipelines/cli/test.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    type: object
+  - name: VSCODE_CLI_DIR
+    type: string
+    default: './'
+  - name: VSCODE_CLI_BINARY_NAME
+    type: string
+  - name: VSCODE_CLI_RUST_CHANNEL
+    type: string
+    default: stable
+
```
Comment (joaomoreno) [inline build/azure-pipelines/cli/test.yml:13]: Should this be configurable?

File: build/azure-pipelines/cli/prepare.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	fs.writeFileSync(path.join(stagingDir, 'package.json'), JSON.stringify({ dependencies: { 'vscode-encrypt': dep } }));
+	cp.execSync('yarn', { cwd: stagingDir, stdio: 'inherit' });
+	const encryptPath = path.join(stagingDir, 'node_modules', 'vscode-encrypt', 'rs-pure');
+
+	addCargoDependency(`vscode-encrypt = { "path" = "${encryptPath.replace(/\\/g, '/')}" }`);
+	enableFeature('vscode-encrypt');
+};
+
+setLauncherEnvironmentVars();
+enableVscodeEncrypt();
```
Comment (joaomoreno) [inline build/azure-pipelines/cli/prepare.ts:86]: These should be wrapped in `if main`.

File: build/azure-pipelines/cli/install-rust.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+# Todo: use 1ES pipeline once extension is installed in ADO
+
+steps:
+  - script: |
+      set -e
+      curl https://sh.rustup.rs -sSf | sh -s -- -y --profile minimal --default-toolchain $RUSTUP_TOOLCHAIN
+      echo "##vso[task.setvariable variable=PATH;]$PATH:$HOME/.cargo/bin"
+    env:
+      RUSTUP_TOOLCHAIN: ${{ parameters.channel }}
+    condition: not(eq(variables['Agent.OS'], 'Windows_NT'))
```
Comment (joaomoreno) [inline build/azure-pipelines/cli/install-rust.yml:18]: We survived so far by having platform specific pipelines for stuff like this. This PR introduces two new ways to distinguish between Posix and Win32. The first I've seen was `VSCODE_IS_POSIX`. Now I see this. Please simplify this and just have platform specific pipelines.

File: build/azure-pipelines/cli/vcpkg-deps.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  - name: targets
+    default: []
+    type: object
+  - name: vcpkgDir
+    type: string
+  - name: targetDirectory
+    type: string
+
+steps:
+  - script: git clone https://github.com/microsoft/vcpkg.git $(Build.ArtifactStagingDirectory)/vcpkg
```
Comment (joaomoreno) [inline build/azure-pipelines/cli/vcpkg-deps.yml:11]: Are we really cloning this bindly? What if this repo gets compromised?

File: build/azure-pipelines/common/createAsset.ts [hunk 6]
```diff
@@ -28,8 +28,10 @@ if (process.argv.length !== 8) {
 	process.exit(-1);
 }
 
+const Ignore = Symbol('Ignore');
```
Comment (joaomoreno) [inline build/azure-pipelines/common/createAsset.ts:31]: Why is this needed? It seems that if we intend to ignore something, we shouldn't call `createAsset` at all for it.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
General comments:

- Injecting targets/artifact names from `product-build.yml` makes the pipelines harder to reason about. Let's be less smart and just describe each platform/arch in the specific `*.yml` files.
- Instead of `build/cli` directory, it's best to have `cli` in the names of the pipelines, as in `build/darwin/cli-compile-darwin`, `build/cli-prepare.yml`, `build/cli-test.yml`, etc.
- Don't use `macos`, but `darwin`.
- The whole `vspkg` usage looks very strange to me. Why is this necessary? Can we abstract this whole thing away in Cargo? If this is only about fetching `openssl`, can't we manually do that instead? How is this `openssl` usage tracked in terms of OSS licensing, component governance, Third-Party-Notices, etc?

General notes:

- It would have been helpful to have reviewed the whitespace changes in a separate PR.

File: build/azure-pipelines/product-build.yml [hunk 7]
```diff
@@ -77,6 +77,10 @@ parameters:
     displayName: "🎯 Web"
     type: boolean
     default: true
+  - name: VSCODE_BUILD_TUNNEL_CLI
+    displayName: "Build Tunnel CLI"
+    type: boolean
+    default: false
```
[the remark this answers is not in the document]
Comment (connor4312) [reply build/azure-pipelines/product-build.yml:83]: Everything is OK if the build it disabled. My thought was to have it "off" initially until Martin's changes to actually _use_ the CLI get in. And having a toggle for a little bit in the Insiders period may be nice so we don't halt a build if something in the CLI gets messed up.

File: build/azure-pipelines/cli/prepare.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	const stagingDir = path.join(tmpdir(), `vscode-encrypt-staging-${Date.now()}`);
+	fs.mkdirSync(stagingDir);
+	fs.writeFileSync(path.join(stagingDir, 'package.json'), JSON.stringify({ dependencies: { 'vscode-encrypt': dep } }));
+	cp.execSync('yarn', { cwd: stagingDir, stdio: 'inherit' });
+	const encryptPath = path.join(stagingDir, 'node_modules', 'vscode-encrypt', 'rs-pure');
+
+	addCargoDependency(`vscode-encrypt = { "path" = "${encryptPath.replace(/\\/g, '/')}" }`);
+	enableFeature('vscode-encrypt');
+};
+
```
[the remark this answers is not in the document]
Comment (connor4312) [reply build/azure-pipelines/cli/prepare.ts:84]: This is really weird. I would be happy to have a better solution. We could have a different Cargo.toml in distro that adds the dependency/feature... I suppose that would work, it's apt to cause conflicts on dependency changes, but that's no different than the package.json...

File: build/azure-pipelines/cli/vcpkg-deps.yml (hunk 5 above)
Comment (connor4312) [reply build/azure-pipelines/cli/vcpkg-deps.yml:11]: I'll have this check out the latest commit on the repo to provide some assurance here.

File: build/azure-pipelines/cli/test.yml (hunk 2 above)
Comment (connor4312) [reply build/azure-pipelines/cli/test.yml:13]: I don't think so. We will never want to publish a build (certainly not from main at least) that's on a non-stable Rust version.

File: build/azure-pipelines/product-build.yml [hunk 9]
```diff
@@ -171,10 +181,108 @@ stages:
...
+                  VSCODE_QUALITY: ${{ variables.VSCODE_QUALITY }}
+                  VSCODE_IS_POSIX: false
+              - template: ./cli/vcpkg-deps.yml
+                parameters:
+                  targets:
+                    - ${{ if eq(parameters.VSCODE_BUILD_WIN32, true) }}:
+                      - x64-windows-static-md
+                    - ${{ if eq(parameters.VSCODE_BUILD_WIN32_ARM64, true) }}:
+                      - arm64-windows-static-md
+                  vcpkgDir: $(Build.SourcesDirectory)/build/azure-pipelines/cli/vcpkg
```
[the remark this answers is not in the document]
Comment (connor4312) [reply build/azure-pipelines/product-build.yml:268]: I just have not yet tested win32, though it should(?) work. That is on my plate for this month, and I can try that in this PR if you want; my goal was to get initial build bits migrated and in without changes to the CLI itself.

File: build/azure-pipelines/product-build.yml [hunk 10]
```diff
@@ -171,10 +181,108 @@ stages:
...
+                parameters:
+                  VSCODE_QUALITY: ${{ variables.VSCODE_QUALITY }}
+                  VSCODE_IS_POSIX: true
+              - template: ./cli/compile-linux.yml
+                parameters:
+                  VSCODE_CLI_DIR: $(Build.SourcesDirectory)/cli
+                  VSCODE_CLI_BINARY_NAME: ${{ variables.VSCODE_CLI_BINARY_NAME }}
+                  VSCODE_CLI_TARGETS:
+                    - ${{ if eq(parameters.VSCODE_BUILD_LINUX_ALPINE, true) }}:
+                      - { target: x86_64-unknown-linux-musl, artifact: vscode_cli_alpine_x64_cli-unsigned }
```
[the remark this answers is not in the document]
Comment (connor4312) [reply build/azure-pipelines/product-build.yml:202]: I've adjusted these slightly, but I still think passing in artifact names is helpful here.

The alternative I see is passing individual flags like `VSCODE_CLI_BUILD_LINUX_X64_MUSL` and doing a big set of if/else's inside the pipeline to pick the target and artifact, but I feel like doing it here is more succinct -- especially since later `product-build-*.yml`'s need to know the artifact names to pull them down as well.

File: build/azure-pipelines/product-build.yml [hunk 11]
```diff
@@ -137,6 +145,8 @@ variables:
     value: true
   - name: Codeql.SkipTaskAutoInjection
     value: true
+  - name: VSCODE_CLI_BINARY_NAME
+    value: code-tunnel
```
[the remark this answers is not in the document]
Comment (joaomoreno) [reply build/azure-pipelines/product-build.yml:145]: But if we inline this today, then we could just rename it inline as well?

File: build/azure-pipelines/product-build.yml [hunk 12]
```diff
@@ -171,10 +177,87 @@ stages:
...
+              vmImage: macOS-latest
+            steps:
+              - template: ./darwin/cli-build-darwin.yml
+                parameters:
+                  VSCODE_QUALITY: ${{ variables.VSCODE_QUALITY }}
+                  VSCODE_CLI_DIR: $(Build.SourcesDirectory)/cli
+                  VSCODE_CLI_BINARY_NAME: ${{ variables.VSCODE_CLI_BINARY_NAME }}
+                  VSCODE_CLI_TARGETS:
+                    - ${{ if eq(parameters.VSCODE_BUILD_MACOS, true) }}:
+                      - { target: x86_64-apple-darwin, artifact: not_yet_signed_vscode_cli_darwin_x64_cli }
```
Comment (joaomoreno) [inline build/azure-pipelines/product-build.yml:236]: `s/not_yet_signed/unsigned/g`, [following prior art](https://github.com/microsoft/vscode/blob/18f743d534ef3f528c5e81e82e695b87c60d2ebf/build/azure-pipelines/darwin/product-build-darwin-universal.yml#L118-L125)

File: build/azure-pipelines/product-build.yml [hunk 13]
```diff
@@ -171,10 +177,87 @@ stages:
...
+      dependsOn: []
+      jobs:
+        - ${{ if or(eq(parameters.VSCODE_BUILD_LINUX, true), eq(parameters.VSCODE_BUILD_LINUX_ALPINE, true)) }}:
+          - job: LinuxX86
+            pool: vscode-1es-linux
+            steps:
+              - template: ./linux/cli-build-linux.yml
+                parameters:
+                  VSCODE_QUALITY: ${{ variables.VSCODE_QUALITY }}
+                  VSCODE_CLI_DIR: $(Build.SourcesDirectory)/cli
```
Comment (joaomoreno) [inline build/azure-pipelines/product-build.yml:191]: Is this ever not going to be `$(Build.SourcesDirectory)/cli`?

File: build/azure-pipelines/product-build.yml [hunk 14]
```diff
@@ -171,10 +177,87 @@ stages:
...
+                  VSCODE_CLI_BINARY_NAME: ${{ variables.VSCODE_CLI_BINARY_NAME }}
+                  VSCODE_CLI_TARGETS:
+                    - ${{ if eq(parameters.VSCODE_BUILD_LINUX_ALPINE, true) }}:
+                      - { target: x86_64-unknown-linux-musl, artifact: vscode_cli_alpine_x64_cli }
+                    - ${{ if eq(parameters.VSCODE_BUILD_LINUX, true) }}:
+                      - { target: x86_64-unknown-linux-gnu, artifact: vscode_cli_linux_x64_cli }
+
```
Comment (joaomoreno) [inline build/azure-pipelines/product-build.yml:198]: Instead of bringing CLI specific knowledge up to `product-build`, could we just forward all `VSCODE_BUILD_LINUX_*` parameters to `cli-build-linux.yml` and have that do the right thing in that pipeline?

File: build/azure-pipelines/cli/prepare.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (joaomoreno) [reply build/azure-pipelines/cli/prepare.ts]: What happened to this code?

**Review by joaomoreno — COMMENTED** [review]
Thanks for simplifying, we're getting there!

File: build/azure-pipelines/cli/cli-compile-and-publish.yml [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+      VSCODE_CLI_AI_ENDPOINT: $(VSCODE_CLI_AI_ENDPOINT)
+      ${{ each pair in parameters.VSCODE_CLI_ENV }}:
+        ${{ pair.key }}: ${{ pair.value }}
+
+  - ${{ if or(contains(parameters.VSCODE_CLI_TARGET, '-windows-'), contains(parameters.VSCODE_CLI_TARGET, '-darwin')) }}:
+    - task: ArchiveFiles@2
+      inputs:
+        ${{ if contains(parameters.VSCODE_CLI_TARGET, '-windows-')}}:
+          rootFolderOrFile: $(Build.SourcesDirectory)/cli/target/${{ parameters.VSCODE_CLI_TARGET }}/release/code-tunnel.exe
+        ${{ if contains(parameters.VSCODE_CLI_TARGET, '-darwin')}}:
```
Comment (joaomoreno) [inline build/azure-pipelines/cli/cli-compile-and-publish.yml:29]: Should this be `elseif`? Or even `else`?

File: build/azure-pipelines/linux/cli-build-linux.yml [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  - name: VSCODE_QUALITY
+    type: string
+  - name: channel
+    type: string
+    default: stable
+
+steps:
+  # inspired by: https://github.com/emk/rust-musl-builder/blob/main/Dockerfile
+  - bash: |
+      sudo apt-get update
```
Comment (joaomoreno) [inline build/azure-pipelines/linux/cli-build-linux.yml:23]: Use `set -e`.

File: build/azure-pipelines/win32/product-build-win32.yml [hunk 17]
```diff
@@ -175,6 +177,35 @@ steps:
...
+    - task: DownloadPipelineArtifact@2
+      inputs:
+        artifact: unsigned_vscode_cli_win32_x64_cli
+        patterns: '**'
+        path: $(Build.ArtifactStagingDirectory)/cli
+      displayName: Download VS Code CLI
+      condition: and(succeeded(), eq(variables['VSCODE_ARCH'], 'x64'))
+
+    - powershell: |
+        . build/azure-pipelines/win32/exec.ps1
```
Comment (joaomoreno) [inline build/azure-pipelines/win32/product-build-win32.yml:198]: Use `$ErrorActionPreference = "Stop"`.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
Last round. Happy to approve and merge this once rest of comments are addressed. ✅

File: build/azure-pipelines/cli/cli-compile-and-publish.yml (hunk 15 above)
Comment (connor4312) [reply build/azure-pipelines/cli/cli-compile-and-publish.yml:29]: I didn't know `else` was a thing in their conditionals!

**Review by joaomoreno — APPROVED** [review]
